_posts/2016-01-26-variance-and-functors.md

```diff
@@ -0,0 +1,309 @@
+a value *from* a `String`, we can write a value *to* a `String`.
+
+```scala
+trait Show[-A] {
+  def show(a: A): String
+}
+```
+
+`Show` is the other side of `Read` - instead of going from a `String` to an `A`, we go from an `A` into
+a `String`. This reversal allows us to define contravariant behavior - if we are asked to provide a way
+to show a `Circle` (`Show[Circle]`), we can give instead a way to show just a `Shape`. This is a valid
+substitution because we can show a `Circle` by throwing away `Circle`-specific information and showing just
+the `Shape` bits. This means that `Show[Shape]` is a subtype of `Show[Circle]`, despite `Circle` being a
+subtype of `Shape`.
+
+In general, we can show (or write) a subtype if we know how to show a supertype by tossing away subtype-specific
+information (an upcast) and showing the remainder. Again, this means `Show[Supertype]` is substitutable, or a
+subtype of, `Show[Subtype]`.
+
+For similar reasons that read-only types can be made covariant, write-only types can be made contravariant.
+
+## Array, again
+`Array`s cannot be made contravariant either. If they were, we could do unsafe reads:
+
+```scala
+val shapes: Array[Shape] = Array.fill(10)(Shape(..), Shape(..))
+val circles: Array[Circle] = shapes // Works only if Array is contravariant
+val circle: Circle = circles(0)
+```
+
+`circle`, having been read from an `Array[Circle]` has type `Circle`. To the compiler this would be fine, but
+at runtime, the underlying `Array[Shape]` may give us a `Shape` that is not a `Circle` and crash the program.
+
+## Contravariant
+Our `Functor` interface made explicit the behavior of covariance - we can define a similar interface that
+captures contravariant behavior. Remember that contravariance is covariance with the direction flipped,
+which translates to the direction of the function being flipped.
+
+```scala
+trait Contravariant[F[_]] {
+  def contramap[A, B](fa: F[A])(f: B => A): F[B]
+}
+```
+
+We can implement an instance for `Show`.
+
+```scala
+val showContravariant: Contravariant[Show] =
+  new Contravariant[Show] {
+    def contramap[A, B](fa: Show[A])(f: B => A): Show[B] =
+      new Show[B] {
+        def show(b: B): String = {
+          val a = f(b)
+          fa.show(toShow)
+        }
+      }
+  }
+```
+
+Here we are saying if we can show an `A`, we can show a `B` by turning a `B` into an `A` before showing it.
+Upcasting is a specific case of this, when `B` is a subtype of `A`.
+
+```scala
+def contraUpcast[F[_], A, B >: A](contra: Contravariant[F], fb: F[B]): F[A] =
+  contra.contramap(fb)((a: A) => a: B)
+```
+
+Going back to `Shape`s and `Circle`s, we can show a `Circle` by upcasting it into a `Shape` and showing that.
+
+# Invariance
+Unannotated type parameters are considered invariant - different instantiations of a type constructor have no
+relationship with one another regardless of what relationships the type parameters may have. Given invariant
```

Comment:
There is a type-relationship equivalent to invariant functors, though it depends on what you mean by 'different'. It's that type equality is congruent.

A = B → F[A] = F[B]

That's not nothing; GADT pattern matching reveals type equalities that only exist in certain blocks.